File: backend/app/utils/ghost_score.py
import math


def calculate_ghost_score(
    ghost_reports: int,
    total_reports: int,
    total_upvotes: int,
    unique_reporters: int,
) -> int:
    if total_reports == 0:
        return 0

    min_sample = 5
    if total_reports < min_sample:
        confidence = total_reports / min_sample
    else:
        confidence = 1.0

    ghost_ratio = ghost_reports / total_reports

    vote_factor = min(total_upvotes / max(total_reports, 1), 3.0) / 3.0

    uniqueness = min(unique_reporters / max(total_reports, 1), 1.0)

    raw = (ghost_ratio * 0.5 + vote_factor * 0.3 + uniqueness * 0.2) * 100

    score = raw * confidence

    bayesian_mean = 40
    bayesian_weight = 3
    smoothed = (score * total_reports + bayesian_mean * bayesian_weight) / (
        total_reports + bayesian_weight
    )

    return max(0, min(100, int(round(smoothed))))


def get_ghost_label(score: int) -> str:
    if score <= 20:
        return "Surprisingly Responsive"
    elif score <= 40:
        return "Mostly Responsive"
    elif score <= 60:
        return "Could Be Better"
    elif score <= 80:
        return "Getting Ghosty"
    else:
        return "Professional Ghost"


def get_ghost_emoji(score: int) -> str:
    if score <= 20:
        return "\U0001f47c"
    elif score <= 40:
        return "\U0001f642"
    elif score <= 60:
        return "\U0001f610"
    elif score <= 80:
        return "\U0001f47b"
    else:
        return "\U0001f480"

File: src/lib/server/ghostScore.ts
export function calculateGhostScore(
    ghostReports: number,
    totalReports: number,
    totalUpvotes: number,
    uniqueReporters: number,
): number {
    if (totalReports === 0) return 0;

    const minSample = 5;
    const confidence = totalReports < minSample ? totalReports / minSample : 1;
    const ghostRatio = ghostReports / totalReports;
    const voteFactor = Math.min(totalUpvotes / Math.max(totalReports, 1), 3) / 3;
    const uniqueness = Math.min(uniqueReporters / Math.max(totalReports, 1), 1);

    const raw = (ghostRatio * 0.5 + voteFactor * 0.3 + uniqueness * 0.2) * 100;
    const score = raw * confidence;

    const bayesianMean = 40;
    const bayesianWeight = 3;
    const smoothed =
        (score * totalReports + bayesianMean * bayesianWeight) / (totalReports + bayesianWeight);

    return Math.max(0, Math.min(100, Math.round(smoothed)));
}

export function getGhostLabel(score: number): string {
    if (score <= 20) return "Surprisingly Responsive";
    if (score <= 40) return "Mostly Responsive";
    if (score <= 60) return "Could Be Better";
    if (score <= 80) return "Getting Ghosty";
    return "Professional Ghost";
}

export function getGhostEmoji(score: number): string {
    if (score <= 20) return "👼";
    if (score <= 40) return "🙂";
    if (score <= 60) return "😐";
    if (score <= 80) return "👻";
    return "💀";
}
